File: Project/Simulator/Abilities/Immortality.java
package Simulator.Abilities;

import Simulator.Tile;
import Simulator.Organisms.Organism;
import Simulator.Organisms.Animals.Animal;
import Simulator.Organisms.Animals.Human;

public class Immortality extends Ability {
    public Immortality(int cooldown, int duration) {
        super(cooldown, duration);
    }

    public Immortality(int cooldown, int duration, int defaultCooldown, int defaultDuration) {
        super(cooldown, duration, 5, 5);
    }

    @Override
    public void effect(Organism user, Organism other) {
        Tile newTile = user.getTile().getRandomFreeNeighbour();
        if (newTile == null) {
            newTile = user.getTile().getRandomNeighbour();
        }
        if (newTile == null) {
            if (other instanceof Animal) {
                ((Animal) other).undoMove();
            }
            return;
        }
        if (user instanceof Human) {
            ((Human) user).move(newTile);
        }
        if (user.getTile().getOrganismCount() > 1) {
            user.collision(user.getTile().getOrganism());
        }
    }

    @Override
    public void effect(Organism user) {
        Tile newTile = user.getTile().getRandomFreeNeighbour();
        if (newTile == null)
            return;
        if (user instanceof Human) {
            ((Human) user).move(newTile);
        }
    }
}

File: Project/Simulator/Abilities/Immortality.py
from Simulator.Organisms.Organism import Organism
from typing import Optional

from .Ability import Ability
from Simulator.Organisms.Animals.Animal import Animal
from Simulator.Organisms.Animals.Human import Human


class Immortality(Ability):
    def __init__(self, cooldown, duration, default_cooldown=-1, default_duration=-1):
        if (default_cooldown == -1) or (default_duration == -1):
            super().__init__(cooldown, duration)
        else:
            super().__init__(cooldown, duration, 5, 5)

    def effect(self, user: Organism, other: Optional[Organism] = None):
        if other is None:
            return
        if user is None:
            return

        if (other is None):
            userTile = user.getTile()
            if (userTile is None):
                return
            newTile = userTile.getRandomFreeNeighbour()
            if newTile is None:
                newTile = userTile.getRandomNeighbour()
            if newTile is None:
                if (isinstance(other, Animal)):
                    other.undoMove()
                return
            if (isinstance(user, Human)):
                user.move(newTile)
            if (userTile.getOrganismCount() > 1):
                user.collision(userTile.getOrganism())
        else:
            newTile = user.getTile()
            if (newTile is None):
                return
            newTile = newTile.getRandomFreeNeighbour()
            if newTile is None:
                return
            if (isinstance(user, Human)):
                user.move(newTile)
